packages/rpc-methods/src/restricted/notify.ts

```diff
@@ -0,0 +1,133 @@
+import {
+  PermissionSpecificationBuilder,
+  PermissionType,
+  RestrictedMethodOptions,
+  ValidPermissionSpecification,
+} from '@metamask/snap-controllers';
+import { NonEmptyArray } from '@metamask/snap-controllers/src/utils';
+import { ethErrors } from 'eth-rpc-errors';
+import { isPlainObject } from '../utils';
+
+const methodName = 'snap_notify';
+
+// Move all the types to a shared place when implementing more notifications
+export enum NotificationType {
+  native = 'native',
+}
+
+export type NotificationArgs = {
+  /**
+   * Enum type to determine notification type. Currently only supports 'native'.
+   */
+  type: NotificationType;
+
+  /**
+   * A message to show on the notification.
+   */
+  message: string;
+};
+
+export type NotifyMethodHooks = {
+  /**
+   * @param snapId - The ID of the Snap that created the notification.
+   * @param args - The notification arguments.
+   */
+  showNotification: (
+    snapId: string,
+    args: NotificationArgs,
+  ) => Promise<{ isRateLimited: boolean }>;
+};
+
+type SpecificationBuilderOptions = {
+  allowedCaveats?: Readonly<NonEmptyArray<string>> | null;
+  methodHooks: NotifyMethodHooks;
+};
+
+type Specification = ValidPermissionSpecification<{
+  permissionType: PermissionType.RestrictedMethod;
+  targetKey: typeof methodName;
+  methodImplementation: ReturnType<typeof getImplementation>;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+}>;
+
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.RestrictedMethod,
+  SpecificationBuilderOptions,
+  Specification
+> = ({ allowedCaveats = null, methodHooks }: SpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.RestrictedMethod,
+    targetKey: methodName,
+    allowedCaveats,
+    methodImplementation: getImplementation(methodHooks),
+  };
+};
+
+export const notifyBuilder = Object.freeze({
+  targetKey: methodName,
+  specificationBuilder,
+  methodHooks: {
+    showNotification: true,
+  },
+} as const);
+
+function getImplementation({ showNotification }: NotifyMethodHooks) {
+  return async function implementation(
+    args: RestrictedMethodOptions<[NotificationArgs]>,
+  ): Promise<null> {
+    const {
+      params,
+      context: { origin },
+    } = args;
+
+    const validatedParams = getValidatedParams(params);
+    const result = await showNotification(origin, validatedParams);
+
+    if (result.isRateLimited) {
+      throw ethErrors.rpc.limitExceeded({
+        message: `Notification type: ${validatedParams.type} is currently rate-limited. Please try again later.`,
```

Comment:
nit:

```suggestion
        message: `Notification type: "${validatedParams.type}" is currently rate-limited. Please try again later.`,
```

Reply:
I almost always quote placeholders in template strings because (1) it'll be obvious add runtime that it's a missing placeholder if something is wrong and the placeholder value is wrong / missing, and (2) I usually want to emphasize the placeholder value.